lib/BitcastUtils.cpp

```diff
@@ -702,4 +702,33 @@ void ConvertInto(Type *Ty, IRBuilder<> &Builder,
   }
 }
 
+bool RemovedCstExprFromFunction(Function *F) {
+  bool Changed = false;
+  for (BasicBlock &BB : *F) {
+    for (Instruction &I : BB) {
+      for (unsigned OperandId = 0; OperandId < I.getNumOperands();
+           OperandId++) {
+        if (auto CE = dyn_cast<ConstantExpr>(I.getOperand(OperandId))) {
+
+          // This case happen in some particular scenario, where it is not
+          // needed to simplify it. Mostly when using global array
+          // variables. In the end llvm managed to deal with it without us
+          // having to simplify it. Trying to simplify would make it very
+          // complicated for the ReplacePointerBitcast pass.
+          if (CE->getOpcode() == Instruction::BitCast &&
+              CE->getOperand(0)
+                  ->getType()
+                  ->getNonOpaquePointerElementType()
+                  ->isStructTy())
+            continue;
```

Comment:
Is this necessary? Does LLVM only handle this oddball case because it is a constantexpr or will it handle the case if there are instructions instead? Ideally this could be removed.

If it has to stay, please a `TODO(#816): remove after final transition` and also guard that the  operand type is a transparent pointer.

Reply:
It is unfortunately still necessary.
here is an example of the problematic input in ReplacePointerBitCast (`NormalizeGlobalVariables/constant_bitcast.cl`):
```
 %1 = bitcast <{ <{ [9 x i32], [8 x i32] }>, [17 x i32], [17 x i32], [17 x i32] }> addrspace(2)* @data to [4 x [17 x i32]] addrspace(2)*
 %3 = getelementptr inbounds [4 x [17 x i32]], [4 x [17 x i32]] addrspace(2)* %1, i32 0, i32 %2, i32 %y
 %4 = load i32, i32 addrspace(2)* %3, align 4
 ```
 At the moment the pass does not support it. But llvm seems to end up simplifying it by itself.